cram_pr2/cram_pr2_environment_manipulation/src/costmaps.lisp

```diff
@@ -0,0 +1,143 @@
+                                              (cl-tf:y (cl-tf:origin handle-pose))
+                                              0))
+         (manipulated-point (cl-tf:make-3d-vector (cl-tf:x (cl-tf:origin manipulated-handle-pose))
+                                                  (cl-tf:y (cl-tf:origin manipulated-handle-pose))
+                                                  0))
+         (V (cl-tf:v- manipulated-point neutral-point))
+         (width (get-width container-name V)))
+    (lambda (x y)
+      (multiple-value-bind (a b c)
+          (line-equation-in-xy neutral-point
+                               manipulated-point)
+        (let* ((P (cl-tf:make-3d-vector x y 0))
+               (dist (line-p-dist a b c P))
+               (dist-p (line-p-dist-point a b c P)))
+          (if (and
+               (< dist (+ (/ width 2) padding))
+               (< (cl-tf:v-norm (cl-tf:v- dist-p neutral-point)) (+ (cl-tf:v-norm V) padding)))
+              0
+              1))))))
+
+(defun make-opened-drawer-side-cost-function (?container-desig arm)
+  (let* ((container-name (string-downcase
+                          (car (alexandria:assoc-value
+                                (desig:description ?container-desig)
+                                :name))))
+         (handle-link (get-handle-link container-name))
+         (handle-pose (get-manipulated-pose (cl-urdf:name handle-link) 0 :relative T))
+         (manipulated-handle-pose (get-manipulated-pose (cl-urdf:name handle-link) 1 :relative T))
+         (neutral-point (cl-tf:make-3d-vector (cl-tf:x (cl-tf:origin handle-pose))
+                                              (cl-tf:y (cl-tf:origin handle-pose))
+                                              0))
+         (manipulated-point (cl-tf:make-3d-vector (cl-tf:x (cl-tf:origin manipulated-handle-pose))
+                                                  (cl-tf:y (cl-tf:origin manipulated-handle-pose))
+                                                  0))
+         (AB (cl-tf:v- manipulated-point neutral-point)))
+    (lambda (x y)
+      (let ((d (cl-tf:z (cl-tf:cross-product AB
+                                             (cl-tf:v- (cl-tf:make-3d-vector x y 0)
+                                                       neutral-point)))))
+        (if (and (< d 0) (eql arm :right))
+            1.0
+            (if (and (> d 0) (eql arm :left))
+                1.0
+                0.0))))))
+
+(defmethod location-costmap:costmap-generator-name->score ((name (eql 'poses-reachable-cost-function))) 10)
+
+(defmethod location-costmap:costmap-generator-name->score ((name (eql 'container-handle-reachable-cost-function))) 10)
+
+(defmethod location-costmap:costmap-generator-name->score ((name (eql 'opened-drawer-cost-function))) 10)
+
+(defmethod location-costmap:costmap-generator-name->score ((name (eql 'opened-drawer-side-cost-function))) 10)
+
+(defparameter *orientation-samples* 3)
```

Comment:
We should propagate these variables into some more top-level stuff. Like location-costmap package. And use a prolog predicate to store the value. Similar to costmap-padding predicate, for example.